numpy/doc/subclassing.py

```diff
@@ -418,6 +418,64 @@ def __array_finalize__(self, obj):
   >>> v.info
   'information'
 
+.. _ndarray-serialization:
+
+Pickling a ``np.ndarray`` subclass
+--------------------------------------------------------
+
+Pickling is the standard process to serialize python objects using the
+``pickle`` package. This serialization capability is usually required
+by the ``multiprocess`` package when parallelizing code.
+
+In order to properly pickle a subclass of ``np.ndarray``, two more methods
+need to be defined in our class definition:
+
+``np.ndarray`` uses the ``__reduce__`` method to tell pickle how the array
+should be serialized. The ``__setstate__`` method sets all the class attributes
+based on the received ``state``.
+
+.. testcode::
+
+  import numpy as np
+
+  class RealisticInfoArray(np.ndarray):
+
+      def __new__(cls, input_array, info=None):
+          # Input array is an already formed ndarray instance
+          # We first cast to be our class type
+          obj = np.asarray(input_array).view(cls)
+          # add the new attribute to the created instance
+          obj.info = info
+          # Finally, we must return the newly created object:
+          return obj
+
+      def __array_finalize__(self, obj):
+          # see InfoArray.__array_finalize__ for comments
+          if obj is None: return
+          self.info = getattr(obj, 'info', None)
+
+      def __reduce__(self):
+          npState = np.ndarray.__reduce__(self)
+          # npState[2] contains the state of the np.ndarray.
+          # We append our attributes to the state
+          myState = npState[2] + (self.info,)
+          # We return the __reduce__ tuple, with our custom state
+          return (npState[0], npState[1], myState)
```

Comment:
There is no guarantee that `ndarray.__reduce__` will continue to return a tuple with exactly three elements or that the last element will be a tuple or indeed that it will even define a `__reduce__` method. All that's guaranteed is that calling pickle on an ndarray will work. I'm not really sure what the best way to do this is... maybe even just something as crude as
```
def _reconstruct_my_array_type(pickled_ndarray, info):
    base = pickle.loads(pickled_ndarray)
    self = base.view(MyArrayType)
    self.info = info
    return self

def __reduce__(self):
    pickled_ndarray = pickle.dumps(np.asarray(self))
    return (_reconstruct_my_array, (pickled_ndarray, self.info))
```

Actually I don't know how the code you wrote is supposed to work anyway, because you're calling `ndarray.__setstate__(arr.__reduce__()[1])` but the second item in a reduce tuple is not supposed to be passed to `__setstate__`...?

(And is there any chance I could convince you not to subclass ndarray at all? Please? AFAICT basically everyone who has ever tried going down this road regrets it, and trying to keep the whole rickety fragile mechanism working is a huge burden on us, and basically in the long run we are trying to work towards a world where no-one does this ever. So if you can possibly avoid it then I strongly recommend it.)